xformers/ops/swiglu.py

```diff
@@ -0,0 +1,109 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All rights reserved.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Optional
+
+import torch
+import torch.nn.functional as F
+from torch import nn
+
+
+class SwiGLUFFN_Reference(nn.Module):
+    """
+    Reference implementation of a SwiGLU module
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        hidden_features: Optional[int] = None,
+        out_features: Optional[int] = None,
+        align_as: int = 8,
+    ) -> None:
+        super().__init__()
+        out_features = out_features or in_features
+        hidden_features = hidden_features or in_features
+        swiglu_hidden_features = int(2 * hidden_features / 3)
+        swiglu_hidden_features = (
+            (swiglu_hidden_features + align_as - 1) // align_as * align_as
+        )
+
+        self.w1 = nn.Linear(in_features, swiglu_hidden_features)
+        self.w2 = nn.Linear(in_features, swiglu_hidden_features)
+        self.w3 = nn.Linear(swiglu_hidden_features, out_features)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x1 = self.w1(x)
+        x2 = self.w2(x)
+        hidden = F.silu(x1) * x2
+        return self.w3(hidden)
+
+    def _ordered_params_for_op(self):
+        """Used for testing - returns ordered arguments for operators"""
+        return [
+            self.w1.weight,
+            self.w1.bias,
+            self.w2.weight,
+            self.w2.bias,
+            self.w3.weight,
+            self.w3.bias,
+        ]
+
+
+def _silu_backward(dy, x):
+    # https://github.com/pytorch/pytorch/blob/563b065f5a4b4055fa6b025c2514b566d5fd9439/aten/src/ATen/native/Activation.cpp#L483
+    sigm = 1 / (1 + torch.exp(-x.float()))
+    return (dy.float() * sigm * (1 + x.float() * (1 - sigm))).to(x.dtype)
+
+
+class SwiGLU_Decomposed(torch.autograd.Function):
+    """
+    This is just an example implementation with all
```

Comment:
It's unclear to me why we would want to expose this in the library -- IMO we should just write reference functional implementations (i.e., not decomposing as `torch.autograd.Function` in the tests and benchmarks.

This is not supposed to be used by the user in any way, right?